data/transactions/logic/pairing.go

```diff
@@ -18,98 +18,766 @@ package logic
-	ret.X.A0 = bytesToBN254Field(b[:32])
-	ret.X.A1 = bytesToBN254Field(b[32:64])
-	ret.Y.A0 = bytesToBN254Field(b[64:96])
-	ret.Y.A1 = bytesToBN254Field(b[96:128])
-	return
-}
+	last := len(cx.stack) - 1
+	prev := last - 1
+	a := cx.stack[prev].Bytes
+	b := cx.stack[last].Bytes
 
-func bytesToBN254G2s(b []byte) (ret []bn254.G2Affine) {
-	for i := 0; i < len(b)/128; i++ {
-		ret = append(ret, bytesToBN254G2(b[(i*128):(i*128+128)]))
+	var res []byte
+	var err error
+	switch fs.field {
+	case BN254g1:
+		res, err = bn254G1Add(a, b)
+	case BN254g2:
+		res, err = bn254G2Add(a, b)
+	case BLS12_381g1:
+		res, err = bls12381G1Add(a, b)
+	case BLS12_381g2:
+		res, err = bls12381G2Add(a, b)
+	default:
+		err = fmt.Errorf("invalid ec_add group %s", group)
 	}
-	return
+	cx.stack[prev].Bytes = res
+	cx.stack = cx.stack[:last]
+	return err
 }
 
-func bN254G1ToBytes(g1 *bn254.G1Affine) (ret []byte) {
-	retX := g1.X.Bytes()
-	retY := g1.Y.Bytes()
-	ret = append(retX[:], retY[:]...)
-	return
-}
+// Input: ToS is a scalar, encoded as an unsigned big-endian, second to top is
+// uncompressed bytes for g1 point
+// Output: Single byte slice on top of stack which contains uncompressed bytes
+// for product of scalar and point
+func opEcScalarMul(cx *EvalContext) error {
+	group := EcGroup(cx.program[cx.pc+1])
+	fs, ok := ecGroupSpecByField(group)
+	if !ok { // no version check yet, both appeared at once
+		return fmt.Errorf("invalid ec_scalar_mul group %s", group)
+	}
 
-func opBn256Add(cx *EvalContext) error {
 	last := len(cx.stack) - 1
 	prev := last - 1
 	aBytes := cx.stack[prev].Bytes
-	bBytes := cx.stack[last].Bytes
-	if len(aBytes) != 64 || len(bBytes) != 64 {
-		return errors.New("expect G1 in 64 bytes")
-	}
-	a := bytesToBN254G1(aBytes)
-	b := bytesToBN254G1(bBytes)
-	res := new(bn254.G1Affine).Add(&a, &b)
-	resBytes := bN254G1ToBytes(res)
+	kBytes := cx.stack[last].Bytes
+	if len(kBytes) > 32 {
+		return fmt.Errorf("ec_scalar_mul scalar len is %d, exceeds 32", len(kBytes))
+	}
+	k := new(big.Int).SetBytes(kBytes)
+
+	var res []byte
+	var err error
+	switch fs.field {
+	case BN254g1:
+		res, err = bn254G1ScalarMul(aBytes, k)
+	case BN254g2:
+		res, err = bn254G2ScalarMul(aBytes, k)
+	case BLS12_381g1:
+		res, err = bls12381G1ScalarMul(aBytes, k)
+	case BLS12_381g2:
+		res, err = bls12381G2ScalarMul(aBytes, k)
+	default:
+		err = fmt.Errorf("invalid ec_scalar_mul group %s", group)
+	}
+
 	cx.stack = cx.stack[:last]
-	cx.stack[prev].Bytes = resBytes
-	return nil
+	cx.stack[prev].Bytes = res
+	return err
 }
 
-func opBn256ScalarMul(cx *EvalContext) error {
+// Input: Two byte slices, top is concatenated uncompressed bytes for k g2 points, and second to top is same for g1
+// Output: Single uint at top representing bool for whether pairing of inputs was identity
+func opEcPairingCheck(cx *EvalContext) error {
+	/*
+	   Q: Should pairing fail is the supplied points are not canonical?
+	   Q: Should pairing fail if there are no inputs?
```

Comment:
Unclear. https://eips.ethereum.org/EIPS/eip-2537 rejects empty inputs.
Using empty inputs is most likely a mistake from the programmers I would guess... Though mathematically, it makes sense to accept empty inputs.

Reply:
Resolved in favor of EIP compatibility (failing on empty).